finance/server/src/main/java/com/google/finapp/SpannerDaoPGImpl.java

```diff
@@ -0,0 +1,367 @@
+      ImmutableList.Builder<TransactionEntry> transactionHistoriesBuilder = ImmutableList.builder();
+      while (resultSet.next()) {
+        transactionHistoriesBuilder.add(
+            TransactionEntry.newBuilder()
+                .setAccountId(ByteString.copyFrom(resultSet.getBytes("AccountId").toByteArray()))
+                .setEventTimestamp(resultSet.getTimestamp("EventTimestamp").toProto())
+                .setIsCredit(resultSet.getBoolean("IsCredit"))
+                .setAmount(resultSet.getBigDecimal("Amount").toString())
+                .build());
+      }
+      return transactionHistoriesBuilder.build();
+    } catch (SpannerException e) {
+      throw Status.fromThrowable(e).asException();
+    }
+  }
+
+  /**
+   * Returns an ImmutableMap of all requested AccountData keyed by account ids.
+   *
+   * @param accountIds account ids for which to return AccountData
+   * @param transaction the transaction context to use
+   * @return mapping of AccountData, keyed by account id
+   * @throws StatusException if any of the requested accountIds is not found, or the associated
+   *     accountStatus is not ACTIVE.
+   */
+  private ImmutableMap<ByteArray, AccountData> readAccountDataForTransfer(
+      Iterable<ByteArray> accountIds, TransactionContext transaction) throws StatusException {
+    KeySet.Builder keySetBuilder = KeySet.newBuilder();
+    for (ByteArray accountId : accountIds) {
+      keySetBuilder.addKey(Key.of(accountId));
+    }
+
+    try (ResultSet resultSet =
+        transaction.read(
+            "Account",
+            keySetBuilder.build(),
+            ImmutableList.of("accountid", "accountstatus", "balance"))) {
+      ImmutableMap.Builder<ByteArray, AccountData> accountDataBuilder = ImmutableMap.builder();
+      while (resultSet.next()) {
+        AccountData accountData = new AccountData();
+        accountData.balance = new BigDecimal(String.valueOf(resultSet.getValue("balance")));
+        accountData.status = AccountStatus.forNumber((int) resultSet.getLong("accountstatus"));
+        accountDataBuilder.put(resultSet.getBytes("accountid"), accountData);
+      }
+      ImmutableMap<ByteArray, AccountData> accountData = accountDataBuilder.build();
+      for (ByteArray accountId : accountIds) {
+        if (!accountData.containsKey(accountId)) {
+          throw Status.INVALID_ARGUMENT
+              .withDescription(String.format("Account not found: %s", accountId.toString()))
+              .asException();
+        } else if (accountData.get(accountId).status != AccountStatus.ACTIVE) {
+          throw Status.INVALID_ARGUMENT
+              .withDescription(
+                  String.format(
+                      "Non-active accounts are not eligible for transfers: %s",
+                      accountId.toString()))
+              .asException();
+        }
+      }
+      return accountData;
+    } catch (SpannerException e) {
```

Comment:
Nit: I'd try to keep the try / catch as specific as possible.

I believe here you need to catch SpannerExceptions only while dealing with the resultSet. After you have the accountData you are not interacting with Spanner anymore.

I'd check other try / catch to see if they're specific enough as well.

Reply:
Not sure I follow. This try/catch block specifically for getting the result set. If this isn't accurate, then the SpannerDaoImpl should be reviewed as well.

Reply:
Feel free to ignore this is minor.

My point is that you could catch after finishing reading the result set.

So I think you could catch before `for (ByteArray accountId : accountIds) {` as the rest doesn't deal with Spanner. Does this make sense to you?

Reply:
ah yes, now I understand.

Reply:
This has been addressed in new commit